### PR TITLE
Make cute dsl mxfp8/nvfp4 quantizer bitwise exact

### PR DESCRIPTION
**Pull request by zianglih**
## 📌 Description
@humansand

We want to make sure FlashInfer quantization backends are bitwise identical with TE style implementation for RL use cases.

Currently cute dsl mxfp8 backend is not bitwise identical on subnormal edge cases, and cute dsl nvfp4 backend does not honor `TRTLLM_DISABLE_FP4_QUANT_FAST_MATH` env var and has fast math always enabled.
- MXFP8
  - Added MXFP8 bitwise-exact reference checks against reference implementation.
  - Fixed MXFP8 CuTe scale conversion edge cases and host scale-output initialization.
- NVFP4
  - Expanded NVFP4 TE-reference exactness coverage to include the CuTe DSL backend.
  - Skipped unsupported CuTe NVFP4 modes in the TE-reference test: per-token activation and 4over6.
  - Made CuTe NVFP4 honor `TRTLLM_DISABLE_FP4_QUANT_FAST_MATH=1` for bitwise-exact math.

## 🔍 Related Issues
- Bitwise per-tensor NVFP4 quantizer for CUDA backend inplemented in:
  - https://github.com/flashinfer-ai/flashinfer/pull/3264
- Bitwise per-token NVFP4 quantizer for CUDA backend implemented in:
  - https://github.com/flashinfer-ai/flashinfer/pull/3027
- Env var `TRTLLM_DISABLE_FP4_QUANT_FAST_MATH ` introduced by:
  - https://github.com/flashinfer-ai/flashinfer/pull/3237

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Optional precise NVFP4 quantization mode (env-controlled) that uses round-to-nearest math, with new RN reciprocal/scale helpers and per-kernel switch for fast-math behavior.

* **Bug Fixes**
  * Fixed an uninitialized buffer in the FP8 quantization CPU path.

* **Tests**
  * Tightened validation to require exact parity, improved mismatch diagnostics, added an extreme-scale MXFP8 test and reference utilities, and backend parametrization.

[Review Change Stack](https://app.coderabbit.ai/change-stack/flashinfer-ai/flashinfer/pull/3387?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by bkryu** [comment]
Hi @zianglih Do you mind updating the PR description and/or link an issue with the motivation, what it is, and why?

Seems like the PR is trying to add a "no-fast-math" path but I'm curious to know why

**Comment by zianglih** [comment]
Hi @bkryu , I have added the description.

File: flashinfer/trace/templates/quantize.py [hunk 1]
```diff
@@ -122,29 +122,97 @@ def _quantize_fp4_block_scale(
...
-    safe = torch.where(block_scale > 0, block_scale, torch.ones_like(block_scale))
-    exp = torch.floor(torch.log2(safe)).to(torch.int64)
-    exp = exp.clamp(-127, 128) + 127
+    # MXFP8 uses UE8M0 scale factors rounded toward +inf. This mirrors the
+    # host reference's FP32 exponent/mantissa handling exactly.
+    scale_bits = block_scale.contiguous().view(torch.int32)
+    exp = torch.bitwise_right_shift(scale_bits, 23) & 0xFF
+    mantissa = scale_bits & 0x7FFFFF
+    round_up = (mantissa > 0) & (exp != 0xFE) & ~((exp == 0) & (mantissa <= 0x400000))
+    exp = exp + round_up.to(torch.int32)
```
[the remark this answers is not in the document]
Comment (zianglih) [reply flashinfer/trace/templates/quantize.py]: Resolved

**Review by aleozlx — APPROVED** [review]
lgtm

**Review by bkryu — APPROVED** [review]

**Comment by zianglih** [comment]
[JIT Unittest (T4)](https://github.com/flashinfer-ai/flashinfer/actions/runs/26473675768/job/77953577319?pr=3387) failure is due to ci environment.
